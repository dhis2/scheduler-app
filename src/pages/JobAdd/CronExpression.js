import React from 'react'
import { shape, string, object } from 'prop-types'
import { InputField, Help } from '@dhis2/ui-core'
import validateCron from '../../services/validate-cron'

export const validate = value => {
    const isString = typeof value === 'string'
    const isFilled = isString && value.length > 0

    if (!isFilled) {
        return new Error('Cron expression is required')
    }

    if (!validateCron(value)) {
        return new Error('Needs to be a valid CRON expression')
    }

    return undefined
}

const CronExpression = ({ input, meta, label }) => {
    const { touched, error } = meta
    const hasError = touched && !!error

    return (
        <div>
            <InputField {...input} label={label} error={hasError} required />
            {hasError && <Help error>{meta.error.message}</Help>}
        </div>
    )
}

CronExpression.propTypes = {
    label: string.isRequired,
    meta: shape({
        error: object,
    }).isRequired,
    input: shape({}).isRequired,
}

CronExpression.validate = validate

export default CronExpression
