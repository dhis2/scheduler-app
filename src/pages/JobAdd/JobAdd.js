import React from 'react'
import { Card, IconInfo16 } from '@dhis2/ui'
import i18n from '@dhis2/d2-i18n'
import { JobAddFormContainer } from '../../components/Forms'
import styles from './JobAdd.module.css'

const infoLink =
    'https://docs.dhis2.org/en/use/user-guides/dhis-core-version-236/maintaining-the-system/scheduling.html'

const JobAdd = () => {
    return (
        <React.Fragment>
            <header className={styles.pageHeader}>
                <h2 className={styles.pageHeaderTitle}>{i18n.t('New Job')}</h2>
            </header>
            <Card className={styles.card}>
                <header className={styles.cardHeader}>
                    <h3 className={styles.cardHeaderTitle}>
                        {i18n.t('Configuration')}
                    </h3>
                    <a
                        href={infoLink}
                        className={styles.cardHeaderLink}
                        target="_blank"
                        rel="noopener noreferrer"
                    >
                        <span className={styles.cardHeaderInfo}>
                            <IconInfo16 />
                        </span>
                        {i18n.t('About job configuration')}
                    </a>
                </header>
                <JobAddFormContainer />
            </Card>
        </React.Fragment>
    )
}

export default JobAdd
