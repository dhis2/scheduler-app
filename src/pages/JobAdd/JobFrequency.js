import React from 'react'
import { shape, string, object } from 'prop-types'
import { SelectField, Help } from '@dhis2/ui-core'

export const validate = value => {
    const isString = typeof value === 'string'
    const isFilled = isString && value.length > 0

    if (!isFilled) {
        return new Error('Job frequency is required')
    }

    return undefined
}

const cronPresets = {
    'Every hour': '0 0 * ? * *',
    'Every day at midnight': '0 0 1 ? * *',
    'Every day at noon': '0 0 12 ? * MON-FRI',
    'Every week': '0 0 3 ? * MON',
}

const JobFrequency = ({ input, meta, label }) => {
    const { touched, error } = meta
    const hasError = touched && !!error

    return (
        <div>
            <SelectField {...input} label={label} error={hasError} required>
                {Object.keys(cronPresets).map(key => (
                    <option key={key} value={cronPresets[key]}>
                        {key}
                    </option>
                ))}
            </SelectField>
            {hasError && <Help error>{meta.error.message}</Help>}
        </div>
    )
}

JobFrequency.propTypes = {
    label: string.isRequired,
    meta: shape({
        error: object,
    }).isRequired,
    input: shape({}).isRequired,
}

JobFrequency.validate = validate

export default JobFrequency
