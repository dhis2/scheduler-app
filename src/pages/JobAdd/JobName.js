import React from 'react'
import { shape, string, object } from 'prop-types'
import { InputField, Help } from '@dhis2/ui-core'

export const validate = value => {
    const isString = typeof value === 'string'
    const isFilled = isString && value.length > 0

    if (!isFilled) {
        return new Error('Job name is required')
    }

    return undefined
}

const JobName = ({ input, meta, label }) => {
    const { touched, error } = meta
    const hasError = touched && !!error

    return (
        <div>
            <InputField {...input} label={label} error={hasError} required />
            {hasError && <Help error>{meta.error.message}</Help>}
        </div>
    )
}

JobName.propTypes = {
    label: string.isRequired,
    meta: shape({
        error: object,
    }).isRequired,
    input: shape({}).isRequired,
}

JobName.validate = validate

export default JobName
