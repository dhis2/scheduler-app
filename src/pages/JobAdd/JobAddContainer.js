import React, { useEffect } from 'react'
import { func, string, bool } from 'prop-types'
import { connect } from 'react-redux'
import { CircularLoader } from '@dhis2/ui-core'
import { actions, selectors } from '../../data/job-types'
import * as rootSelectors from '../../rootSelectors'
import { AbsoluteCenter } from '../../components/AbsoluteCenter'
import JobAdd from './JobAdd'

export const UnconnectedJobAddContainer = ({
    didFetch,
    errorMessage,
    fetchJobTypesIfNeeded,
}) => {
    useEffect(() => {
        fetchJobTypesIfNeeded()
    }, [fetchJobTypesIfNeeded])

    // Show spinner when the options haven't loaded yet
    if (!didFetch) {
        return (
            <AbsoluteCenter vertical>
                <CircularLoader />
                Loading form options
            </AbsoluteCenter>
        )
    }

    if (errorMessage) {
        return <span>{errorMessage}</span>
    }

    return <JobAdd />
}

UnconnectedJobAddContainer.propTypes = {
    didFetch: bool.isRequired,
    isFetching: bool.isRequired,
    errorMessage: string.isRequired,
    createJob: func.isRequired,
    fetchJobTypesIfNeeded: func.isRequired,
}

const mapStateToProps = state => {
    const jobTypes = rootSelectors.getJobTypes(state)

    return {
        didFetch: selectors.getDidFetch(jobTypes),
        isFetching: selectors.getIsFetching(jobTypes),
        errorMessage: selectors.getErrorMessage(jobTypes),
    }
}

const mapDispatchToProps = {
    fetchJobTypesIfNeeded: actions.fetchJobTypesIfNeeded,
}

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(UnconnectedJobAddContainer)
