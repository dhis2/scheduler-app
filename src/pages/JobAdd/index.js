import JobAdd from './JobAdd'

export default JobAdd
