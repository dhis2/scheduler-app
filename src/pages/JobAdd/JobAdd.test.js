import React from 'react'
import { shallow } from 'enzyme'
import JobAdd from './JobAdd'

describe('<JobAdd>', () => {
    it('renders correctly', () => {
        const wrapper = shallow(<JobAdd />)

        expect(wrapper).toMatchSnapshot()
    })
})
