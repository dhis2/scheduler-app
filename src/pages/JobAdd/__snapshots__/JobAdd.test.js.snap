// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<JobAdd> renders correctly 1`] = `
<Fragment>
  <Link
    to="/"
  >
    Back to all jobs
  </Link>
  <Title
    priority={2}
  >
    New Job
  </Title>
  <Card>
    <Title
      priority={3}
    >
      Configuration
    </Title>
    <Info />
    <a
      href="https://docs.dhis2.org/master/en/user/html/dataAdmin_scheduling.html#dataAdmin_scheduling_config"
    >
      About job configuration
    </a>
    <InputField
      label="Name"
      onBlur={[Function]}
      onFocus={[Function]}
      type="text"
    />
    <SelectField
      label="Frequency"
    >
      <option
        value="0"
      >
        Continuous execution
      </option>
    </SelectField>
    <SelectField
      label="Job type"
    >
      <option
        value="0"
      >
        Monitoring
      </option>
    </SelectField>
    <Button
      primary={true}
      type="button"
    >
      Save job
    </Button>
    <Button
      type="button"
    >
      Cancel
    </Button>
  </Card>
</Fragment>
`;
